llmfoundry/data/finetuning/dataloader.py

```diff
@@ -199,28 +159,92 @@ def build_finetuning_dataloader(cfg: DictConfig,
-            persistent_workers=cfg.get('persistent_workers', True),
-            timeout=cfg.get('timeout', 0),
-        )
+            shuffle=cfg.dataset.shuffle)
+    else:
+        # Build streaming dataset
+        dataset = _build_streaming_dataset(cfg.dataset, tokenizer, device_batch_size)
+    
+    assert dataset is not None
+    dl = DataLoader(
+        dataset,
+        collate_fn=collate_fn,
+        batch_size=dataloader_batch_size,
+        drop_last=cfg.drop_last,
+        sampler=sampler,
+        num_workers=cfg.num_workers,
+        pin_memory=cfg.get('pin_memory', True),
+        prefetch_factor=cfg.get('prefetch_factor', 2),
+        persistent_workers=cfg.get('persistent_workers', True),
+        timeout=cfg.get('timeout', 0),
+    )
 
     token_counting_func = get_tokens_per_batch_func(
         pad_token_id=tokenizer.pad_token_id)
 
     return DataSpec(dataloader=dl, get_num_tokens_in_batch=token_counting_func)
 
+def _build_preprocessing_function(dataset_name_or_path: str, proto_preprocessing_fn: Optional[Union[dict, DictConfig, str]]):
+    if isinstance(proto_preprocessing_fn, dict) or isinstance(
+            proto_preprocessing_fn, DictConfig):
+        return dataset_constructor.get_preprocessing_fn_from_dict(
+            proto_preprocessing_fn)
+    return dataset_constructor.get_preprocessing_fn_from_str(
+            proto_preprocessing_fn, dataset_name_or_path)
+
+def _build_hf_dataset(hf_name: str, split: Optional[str], safe_load: bool, max_seq_len: int, 
+                      preprocessing_fn: Optional[Callable[[dict[str, Any]], 
+                                                          dict[str,str]]], hf_kwargs: dict) -> Union[hf_datasets.DatasetDict, hf_datasets.Dataset, hf_datasets.IterableDatasetDict, hf_datasets.IterableDataset]:
+    backend, _, _ = parse_uri(hf_name)
+    if backend not in ['', None]:
+        # Download dataset from remote object store. 
+        if split is None:
+            raise ValueError(
+                'When using a HuggingFace dataset from a URL, you must set the ' + \
+                '`split` key in the dataset config.'
+            )
+        hf_name = _download_remote_dataset(hf_name, split)
+    elif not os.path.exists(hf_name) and safe_load: # hf_name is not a local directory and safe_load is True
+        # Download dataset from huggingface hub with restrictions.
+        token = hf_kwargs.get('token', None)
+        hf_name = _safe_download_hf_dataset(hf_name, token)
+    # Build HF dataset
+    return dataset_constructor.build_from_hf(
+        dataset_name_or_path=hf_name,
+        split=split,
+        max_seq_len=cfg.dataset.max_seq_len,
+        preprocessing_fn=preprocessing_fn,
+        tokenizer=tokenizer,
+        hf_kwargs=hf_kwargs
+    )
+
+
+def _build_streaming_dataset(dataset_cfg: DictConfig, tokenizer: PreTrainedTokenizerBase, device_batch_size: int) -> StreamingFinetuningDataset:
+    return dataset_constructor.build_from_streaming(
```

Comment:
FYI: Streaming dataset added `allow_unsafe_types` flag as part of PR https://github.com/mosaicml/streaming/pull/519. Soon, streaming dataset will have a new patch release version with this change. @knighton created the PR https://github.com/mosaicml/llm-foundry/pull/785 to add this flag.